### ISSUE TITLE
3x slowdown after thread affinity change

### ISSUE THREAD

**Issue by danadam** [ggreer/the_silver_searcher#767]
System: Ubuntu 12.04, 24 GB RAM, SSD

CPU:

    vendor_id       : GenuineIntel
    cpu family      : 6
    model           : 44
    model name      : Intel(R) Core(TM) i7 CPU         970  @ 3.20GHz

In a directory with:

    ]$ find -type f | wc -l
    18795

The results of version from before the change:

    43ab7fa0b0 (2014-12-25 09:18:35 -0800) Geoff Greer - Merge pull request #545 from blueyed/add-debug-load-ignore
    ]$ for ((i = 0; i < 3; i++)); do time ag "hello" >/dev/null; sleep 5; done

    real    0m0.587s
    user    0m0.780s
    sys     0m0.880s

    real    0m0.585s
    user    0m0.768s
    sys     0m0.884s

    real    0m0.566s
    user    0m0.812s
    sys     0m0.816s

The results of version from after the change:

    75a260565587 (2014-12-25 09:30:25 -0800) Geoff Greer - Merge pull request #565 from ggreer/thread_affinity
    ]$ for ((i = 0; i < 3; i++)); do time ag "hello" >/dev/null; sleep 5; done

    real    0m1.549s
    user    0m1.304s
    sys     0m1.492s

    real    0m1.576s
    user    0m1.164s
    sys     0m1.388s

    real    0m1.438s
    user    0m1.220s
    sys     0m1.492s

Beginning of the debug:

    DEBUG: Found user's home dir: /home/xxxx
    DEBUG: Skipping ignore file /home/xxxx/.agignore: not readable
    DEBUG: Found user's global Git excludesfile: /home/xxxx/.gitignore
    DEBUG: Loading ignore file /home/xxxx/.gitignore.
    DEBUG: added ignore pattern swp to root ignores
    DEBUG: added ignore pattern swo to root ignores
    DEBUG: Query is hello
    DEBUG: PCRE Version: 8.12 2011-01-15
    DEBUG: Using 11 workers
    DEBUG: Thread 0 set to CPU 0
    DEBUG: Worker 0 started
    DEBUG: Worker 1 started
    DEBUG: Thread 1 set to CPU 1
    DEBUG: Worker 2 started
    DEBUG: Thread 2 set to CPU 2
    DEBUG: Worker 3 started
    DEBUG: Thread 3 set to CPU 3
    DEBUG: Thread 4 set to CPU 4
    DEBUG: Worker 5 started
    DEBUG: Thread 5 set to CPU 5
    DEBUG: Worker 6 started
    DEBUG: Thread 6 set to CPU 6
    DEBUG: Worker 7 started
    DEBUG: Thread 7 set to CPU 7
    DEBUG: Thread 8 set to CPU 8
    DEBUG: Worker 8 started
    DEBUG: Thread 9 set to CPU 9
    DEBUG: Worker 10 started
    DEBUG: Worker 9 started
    DEBUG: Thread 10 set to CPU 10
    DEBUG: searching path . for hello
    DEBUG: Skipping ignore file ./.agignore: not readable
    DEBUG: Loading ignore file ./.gitignore.
    DEBUG: added ignore pattern *.pro to
    ...

**Comment by ggreer** [comment]
Can you try the latest ag with `--affinity` and `--noaffinity`? That way I can be sure it's thread affinity causing the problem and not something else.

**Comment by danadam** [comment]
4914996e4640 (2015-10-04 23:57:21 -0700) Geoff Greer - Merge branch 'natemara-patch-1'                                                                                                     

    ]$ for ((i = 0; i < 3; i++)); do time ag --affinity "hello" >/dev/null; sleep 5; done

    real    0m1.729s
    user    0m1.212s
    sys     0m1.520s

    real    0m1.481s
    user    0m1.288s
    sys     0m1.512s

    real    0m1.522s
    user    0m1.204s
    sys     0m1.532s

    ]$ for ((i = 0; i < 3; i++)); do time ag --noaffinity "hello" >/dev/null; sleep 5; done

    real    0m0.571s
    user    0m0.764s
    sys     0m0.824s

    real    0m0.569s
    user    0m0.788s
    sys     0m0.804s

    real    0m0.513s
    user    0m0.608s
    sys     0m0.864s

**Comment by mitchblank** [comment]
Yeah, on a fairly high-core box (quad-socket E5-4650 -- so that's 32 cores, or 64 counting hyperthreading; CentOS 7.1; kernel 3.10.0-229.14.1.el7.x86_64; glibc-2.17-78.el7; way more than enough RAM to avoid any I/O in this test; filesystem mounted `noatime`) I see `--noaffinity` win by roughly 2x:

```
$ time ag --stats --affinity  sdkfjsdf
0 matches
0 files contained matches
15166 files searched
152809703 bytes searched
2.837769 seconds
1.116/2.270/118.94% user/kern, 0m2.847s elapsed

$ time ag --stats --noaffinity  sdkfjsdf
0 matches
0 files contained matches
15166 files searched
152809703 bytes searched
1.581965 seconds
1.116/1.939/192.12% user/kern, 0m1.590s elapsed
```

both of these were with warm cache of course.  Just eyeballing it, w/o affinity the range seems from about 1.4s..1.9s.  With affinity is always slower; about 2.5s..2.9s.

One thing to note is that this is a shared machine; although there is 90% CPU idle when I ran the tests there are other work being done.

One thing that's notable is that in neither case did it come close to being CPU bound on this box.  We're not doing I/O; we're just limited by how fast the kernel can hand us files.  Probably the affinity is forcing it onto more CPUs than are actually helpful for the work to do, so it's just causing the syscalls to be made from places with colder caches.  Here is the first lines of the output of `strace -f -c`:

```
NO AFFIINITY:

% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 99.44  605.457019        3593    168487     34051 futex
  0.12    0.757713          39     19573           munmap
  0.12    0.711869          19     36811     16921 open
  0.08    0.473843          24     19625           mmap
  0.07    0.455419          20     23277           close
  0.07    0.450265          23     19557           madvise
  0.07    0.431047          22     19888           fstat
  0.01    0.049841          15      3414         9 stat
  0.01    0.040268           6      6774           getdents
  0.00    0.027862           8      3385           openat
  0.00    0.008689        2896         3         1 wait4

WITH AFFINITY:

% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 99.30  657.745683        3896    168834     34934 futex
  0.15    1.005291          27     36811     16921 open
  0.15    0.980085          50     19573           munmap
  0.10    0.648146          33     19625           mmap
  0.10    0.631886          27     23277           close
  0.09    0.603664          31     19557           madvise
  0.09    0.599786          30     19888           fstat
  0.01    0.067706          20      3414         9 stat
  0.01    0.063536           9      6774           getdents
  0.01    0.043036          13      3385           openat
```

You can see that just about every syscall ran slower when we forced affinity onto the scheduler.  The cost of a "open+fstat+mmap+close+madvise+munmap" went from 147µs to 198µs.  Since the mean file size is only 10K that OS cost is the limiting factor: we have 15..20ns of OS latency for every byte we search.

(Aside: I didn't try it, but I bet `madvise` is costing more than it gives in this case; you'd probably get some benefit if you only called it for large files)

That's not super surprising, really.  There were probably lots of times when the scheduler would have wanted to wake up a thread on a CPU that was already cache-hot to this process but instead had to run it somewhere not as optimal.  Worse, since there are other processes on the box it could be that waking up that thread has to wait behind some other process running on it!

Also notable is the TLB shootdown cost of `munmap` which is always substantial in multithreaded programs as you scale up the number of CPUs.  If the searching (and thus the mmap/munmap) could be done in separate processes instead of threads that part might be quicker... of course, it makes distributing work and retrieving results quite a lot more complicated.

tl;dr -- if the CPU cost of searching isn't your dominant cost then affinity is a loss, especially as the number of cores increases.

**Comment by Jecvay** [comment]
```
[Dev18A]$ time ag --cc --affinity --workers 14 friend > /dev/null                                                                                                                                                                                     
ag --cc --affinity --workers 14 friend > /dev/null  1.83s user 92.31s system 1161% cpu 8.102 total
[Dev18A]$ time ag --cc --noaffinity --workers 14 friend > /dev/null                                                                                                                                                                                   
ag --cc --noaffinity --workers 14 friend > /dev/null  1.74s user 91.15s system 1164% cpu 7.976 total
[Dev18A]$ time ag --cc --affinity --workers 1 friend > /dev/null                                                                                                                                                                                      
ag --cc --affinity --workers 1 friend > /dev/null  1.48s user 0.74s system 115% cpu 1.927 total
[Dev18A]$ time ag --cc --noaffinity --workers 1 friend > /dev/null                                                                                                                                                                                    
ag --cc --noaffinity --workers 1 friend > /dev/null  1.43s user 0.76s system 114% cpu 1.916 total
```

this is really curious...
